### PR TITLE
feat(skills): hierarchical skill optimization and search command

### PR DESCRIPTION
**Pull request by dumko2001**
## Description

This PR implements hierarchical skill optimization to solve the context-pollution problem raised in #82. Instead of loading all 40+ API service skills at once, we now use a "Discovery-First" architecture inspired by Claude and GitHub Copilot.

Key changes:
- **Hierarchical Structure:** High-level Personas and Recipes remain in the root `skills/` directory, while granular technical references for individual services/helpers are moved to `skills/references/`.
- **`gws skills search <query>`:** A new command for semantic/keyword discovery of tools. This allows agents to find the right skill without pre-loading the entire library.
- **Enhanced `gws-shared`:** Restored all safety rails (zsh expansion, JSON quoting) and added a "Discovery & Search" section.
- **Automated Generation:** Refactored `generate-skills` to handle the new hierarchy and update all internal Markdown cross-references automatically.
- **Cleanup:** Removed 188 committed Markdown files (6,000+ lines of bloat) from the repository.

**Dry Run Output:**
(Since `gws skills search` is a local discovery tool, it does not produce a JSON API request. Here is the command output instead):
```bash
$ gws skills search "email"
Searching for skills matching "email"...

[Service] gws-gmail - Send, read, and manage email
  Reference: skills/references/gws-gmail/SKILL.md

[Recipe] recipe-label-and-archive-emails - Label and Archive Gmail Threads
  Description: Apply Gmail labels to matching messages and archive them to keep your inbox clean.
  Skill: skills/recipe-label-and-archive-emails/SKILL.md
...
Found 8 matching skills.
```

## Checklist:

- [x] My code follows the `AGENTS.md` guidelines (no generated `google-*` crates).
- [x] I have run `cargo fmt --all` to format the code perfectly.
- [x] I have run `cargo clippy -- -D warnings` and resolved all warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have provided a Changeset file (e.g. via `pnpx changeset`) to document my changes.

### CODE REVIEW

File: check_links.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import os
+import re
+from pathlib import Path
+
+def check_links():
+    basedir = Path("skills").resolve()
+    docsdir = Path("docs").resolve()
+    
+    # regex to find markdown links: [text](path)
+    link_regex = re.compile(r'\[.*?\]\(([^http].*?)\)')
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply check_links.py]: Fixed the link regex to match any path and filtered out HTTP(S) links in the loop as suggested.

File: .gitignore [hunk 2]
```diff
@@ -47,4 +47,4 @@ docs/plans/
 
 # Generated
 demo.mp4
-download.html
\ No newline at end of file
+download.htmlskills/
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply .gitignore]: Fixed the formatting issue in .gitignore. The skills/ directory is now on its own line.

File: check_links.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+    
+    broken_links = 0
+    total_links = 0
+    
+    files_to_check = list(basedir.rglob("*.md")) + list(docsdir.rglob("*.md"))
+    
+    for filepath in files_to_check:
+        with open(filepath, 'r') as f:
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply check_links.py]: Added encoding='utf-8' to the open() call.

File: check_links.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dumko2001) [reply check_links.py]: Implemented explicit error propagation for YAML parsing in both  and . This ensures that any malformed registry files are reported as hard errors instead of failing silently.

File: src/skills_search.rs [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                || r.title.to_lowercase().contains(&query)
+                || r.description.to_lowercase().contains(&query)
+            {
+                println!("[Recipe] recipe-{} - {}", r.name, r.title);
+                println!("  Description: {}", r.description);
+                println!("  Skill: skills/recipe-{}/SKILL.md\n", r.name);
+                results += 1;
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs]: Fixed. Replaced  with explicit error propagation using  to ensure YAML parsing errors are reported.

File: src/skills_search.rs [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+struct RecipeRegistry {
+    recipes: Vec<RecipeEntry>,
+}
+
+#[derive(Deserialize)]
+struct RecipeEntry {
+    name: String,
+    title: String,
+    description: String,
+}
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs]: Great suggestion. Centralized , , , and  into a new  module. Both  and  now import these definitions to ensure consistency and maintainability.

File: src/skills_search.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs]: Fixed. Replaced `if let Ok` with explicit error propagation using `map_err` to ensure YAML parsing errors are reported.

File: src/skills_search.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs]: Great suggestion. Centralized `PersonaRegistry`, `PersonaEntry`, `RecipeRegistry`, and `RecipeEntry` into a new `src/registry.rs` module. Both `skills_search.rs` and `generate_skills.rs` now import these definitions to ensure consistency and maintainability.

File: src/skills_search.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    let query = args
+        .get(1)
+        .ok_or_else(|| {
+            GwsError::Validation(
+                "No search query provided. Usage: gws skills search <query>".to_string(),
+            )
+        })?
+        .to_lowercase();
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs]: Fixed in latest commit. Now joining all arguments following the search command to support multi-word queries without quotes.

File: src/skills_search.rs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    }
+
+    if results == 0 {
+        println!("No matching skills found.");
+    } else {
+        println!("Found {} matching skills.", results);
+    }
+
+    Ok(())
+}
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs:136]: Fixed in latest commit. Added dynamic search for helper commands across all services by inspecting the clap command tree.

File: src/skills_search.rs [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+        return Err(GwsError::Validation(
+            "Usage: gws skills search <query>".to_string(),
+        ));
+    }
+
+    if args.len() < 2 {
+        return Err(GwsError::Validation(
+            "No search query provided. Usage: gws skills search <query>".to_string(),
+        ));
+    }
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs:63]: Fixed. Added explicit `--help`/`-h` handling in `handle_skills_command`: both `gws skills --help` and `gws skills search --help` now print a dedicated help screen (`print_skills_help()`) and return `Ok(())` instead of treating the flag as a search query. This avoids a full clap refactor—consistent with the codebase's manual dispatch pattern for all other top-level commands—while eliminating the incorrect behavior.

**Comment by dumko2001** [comment]
## Summary of all Gemini review fixes

Addressed all reviewer feedback across the 4 review rounds:

**Round 1** (on initial commit):
- ✅ `check_links.py`: Fixed link regex to correctly resolve relative paths (`[text](path)`) and skip HTTP(S) links in-loop
- ✅ `check_links.py`: Added `encoding='utf-8'` to `open()` to prevent `UnicodeDecodeError` on Windows
- ✅ `.gitignore`: Fixed missing newline before `skills/` entry so it is correctly recognized as its own pattern

**Round 2** (error handling):
- ✅ `skills_search.rs` / `generate_skills.rs`: Replaced silent `if let Ok` YAML parsing with explicit `.map_err(...)?` propagation so malformed registry files surface as real errors

**Round 3** (code duplication):
- ✅ Centralized `PersonaRegistry`, `PersonaEntry`, `RecipeRegistry`, `RecipeEntry`, `PERSONAS_YAML`, and `RECIPES_YAML` into a new `src/registry.rs` module; both `generate_skills.rs` and `skills_search.rs` now import from it

**Round 4** (multi-word queries):
- ✅ `skills_search.rs`: Replaced `.get(1)` with `args[1..].join(" ")` so `gws skills search send email` works without quotes

**Round 5** (helper skills missing from search):
- ✅ Added dynamic helper search: iterates all services, calls `helpers::get_helper` + `inject_commands` with a minimal `RestDescription`, and matches `+cmd` subcommands by name and description

**Round 6** (latest — `--help` not handled):
- ✅ Added `print_skills_help()` with proper USAGE / DESCRIPTION / ARGUMENTS / EXAMPLES output; both `gws skills --help` and `gws skills search --help` now print it and return `Ok(())` instead of treating the flag as a query

All 630 tests pass. `cargo clippy -- -D warnings` is clean.

File: src/skills_search.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs]: One more bug found during deeper architectural audit: the previous `args[1..].join(" ")` approach required the multi-word query to appear as a **contiguous substring** in the field text. So `gws skills search send email` built the query string `"send email"` and failed to match `"Gmail: Send an email"` (words present but not adjacent). Fixed in the latest commit with a **token-AND** approach: the query is split into individual tokens and all tokens must appear anywhere in the combined fields — matching how natural-language search is expected to work.

File: check_links.py [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+    total_links = 0
+    
+    files_to_check = list(basedir.rglob("*.md")) + list(docsdir.rglob("*.md"))
+    
+    for filepath in files_to_check:
+        with open(filepath, 'r', encoding='utf-8') as f:
+            content = f.read()
+            
+        for match in link_regex.finditer(content):
+            link_path = match.group(1).strip()
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply check_links.py]: Fixed — applied the suggested  to strip the optional title from captured link paths. commit 20e4c259.

File: src/generate_skills.rs [hunk 10]
```diff
@@ -96,13 +63,44 @@ pub async fn handle_generate_skills(args: &[String]) -> Result<(), GwsError> {
...
+                            matches_helper = true;
+                            break;
+                        }
+                    }
+                }
+                if !matches_helper {
+                    continue;
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/generate_skills.rs:92]: Fixed — replaced the build_cli() helper-filter loop in generate_skills.rs with a simple HELPERS registry lookup. The new code filters by service_alias and checks if the formatted helper name contains the filter string, with no clap::Command construction needed. commit 20e4c259.

File: src/skills_search.rs [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+                        println!(
+                            "  Reference: skills/references/{}/SKILL.md\n",
+                            full_helper_name
+                        );
+                        results += 1;
+                    }
+                }
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs]: Fixed — replaced the inject_commands() clap loop with a direct iteration over the HELPERS static registry in registry.rs. Helper metadata (service_alias, name, description) is now a compile-time constant, eliminating runtime CLI construction entirely. commit 20e4c259.

File: src/skills_search.rs [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    println!("    <query>    Keywords to search for (multi-word queries are supported)");
+    println!();
+    println!("EXAMPLES:");
+    println!("    gws skills search email");
+    println!("    gws skills search \"send email\"");
+    println!("    gws skills search upload file");
+}
+
+/// Entry point for `gws skills search <query>`.
+pub async fn handle_skills_command(args: &[String]) -> Result<(), GwsError> {
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/skills_search.rs:46]: Added — extracted the token-matching closure into a testable pub(crate) token_matches() function and added a #[cfg(test)] module with 7 unit tests covering: single token match/no-match, multi-token AND matching, partial failure, case insensitivity, cross-field matching, and empty-token edge case. commit 20e4c259.

File: src/generate_skills.rs [hunk 13]
```diff
@@ -156,7 +144,7 @@ pub async fn handle_generate_skills(args: &[String]) -> Result<(), GwsError> {
         if emit_service {
             let service_md =
                 render_service_skill(alias, entry, &helpers, &resources, &product_name, &doc);
-            write_skill(output_path, &skill_name, &service_md)?;
+            write_skill(output_path, &skill_name, &service_md, true)?;
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/generate_skills.rs:147]: Declining this suggestion — the `true` flag is intentional and correct. `write_skill(..., true)` routes output to `skills/references/{name}/SKILL.md`, which is the core feature of this PR (hierarchical structure). Changing it to `false` would write service skills back into `skills/{name}/SKILL.md`, reverting the hierarchical layout entirely. The `?` is also kept to propagate I/O errors rather than silently swallowing them.

File: src/generate_skills.rs [hunk 14]
```diff
@@ -184,7 +172,7 @@ pub async fn handle_generate_skills(args: &[String]) -> Result<(), GwsError> {
                 let about_clean = about_raw.strip_prefix("[Helper] ").unwrap_or(&about_raw);
                 let helper_md =
                     render_helper_skill(alias, helper_name, helper, entry, &product_name);
-                write_skill(output_path, &helper_skill_name, &helper_md)?;
+                write_skill(output_path, &helper_skill_name, &helper_md, true)?;
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/generate_skills.rs:175]: Same as above — declining this suggestion. Helper skills are also part of the hierarchical structure and must go into `skills/references/gws-{service}-{helper}/SKILL.md`. Using `is_reference: false` here would break that. The `?` stays to propagate errors correctly.

File: src/generate_skills.rs [hunk 15]
```diff
@@ -755,7 +775,7 @@ fn render_persona_skill(persona: &PersonaEntry) -> String {
     let required_skills = persona
         .services
         .iter()
-        .map(|s| format!("\"gws-{s}\""))
+        .map(|s| format!("\"references/gws-{s}\""))
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/generate_skills.rs:778]: This is already implemented — the current code at line 778 already reads `.map(|s| format\!("\"references/gws-{s}\""))`. The diff you're reviewing shows the old line (gws-{s}) removed and the new line (references/gws-{s}) added as part of this PR.

File: src/generate_skills.rs [hunk 16]
```diff
@@ -825,7 +845,7 @@ fn render_recipe_skill(recipe: &RecipeEntry) -> String {
     let required_skills = recipe
         .services
         .iter()
-        .map(|s| format!("\"gws-{s}\""))
+        .map(|s| format!("\"references/gws-{s}\""))
```
[the remark this answers is not in the document]
Comment (dumko2001) [reply src/generate_skills.rs:848]: Same as above — already implemented. Line 848 already reads `.map(|s| format\!("\"references/gws-{s}\""))`. The suggestion matches what this PR already puts in place.
